state/validatorsInfoMap.go

```diff
@@ -0,0 +1,224 @@
+package state
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+)
+
+type shardValidatorsInfoMap struct {
+	mutex      sync.RWMutex
+	valInfoMap map[uint32][]ValidatorInfoHandler
+}
+
+// NewShardValidatorsInfoMap creates an instance of shardValidatorsInfoMap which manages a
+// <shardID, validatorsInfo> map internally
+func NewShardValidatorsInfoMap(numOfShards uint32) *shardValidatorsInfoMap {
+	return &shardValidatorsInfoMap{
+		mutex:      sync.RWMutex{},
+		valInfoMap: make(map[uint32][]ValidatorInfoHandler, numOfShards),
+	}
+}
+
+// TODO: Delete this once map[uint32][]*ValidatorInfo is completely replaced with new interface
+
+// CreateShardValidatorsMap creates an instance of shardValidatorsInfoMap which manages a shard validator
+// info map internally.
+func CreateShardValidatorsMap(input map[uint32][]*ValidatorInfo) *shardValidatorsInfoMap {
+	ret := &shardValidatorsInfoMap{valInfoMap: make(map[uint32][]ValidatorInfoHandler, len(input))}
+
+	for shardID, valInShard := range input {
+		for _, val := range valInShard {
+			ret.valInfoMap[shardID] = append(ret.valInfoMap[shardID], val)
+		}
+	}
+
+	return ret
+}
+
+// GetAllValidatorsInfo returns a []ValidatorInfoHandler copy with validators from all shards.
+func (vi *shardValidatorsInfoMap) GetAllValidatorsInfo() []ValidatorInfoHandler {
+	ret := make([]ValidatorInfoHandler, 0)
+
+	vi.mutex.RLock()
+	defer vi.mutex.RUnlock()
+
+	for _, validatorsInShard := range vi.valInfoMap {
+		validatorsCopy := make([]ValidatorInfoHandler, len(validatorsInShard))
+		copy(validatorsCopy, validatorsInShard)
+		ret = append(ret, validatorsCopy...)
+	}
+
+	return ret
+}
+
+// GetShardValidatorsInfoMap returns a <shard, ValidatorInfoHandler> map copy of internally stored data
+func (vi *shardValidatorsInfoMap) GetShardValidatorsInfoMap() map[uint32][]ValidatorInfoHandler {
+	ret := make(map[uint32][]ValidatorInfoHandler, len(vi.valInfoMap))
```

Comment:
len not required

Reply:
Same comment as above. This link also states something about it:

https://stackoverflow.com/questions/53388332/memory-efficient-implementation-of-go-map
> Go maps are optimized for integer keys. Optimize the map allocation by giving the exact map size as a hint.